setup.py

```diff
@@ -26,6 +33,29 @@
     package_data.setdefault('setuptools.command', []).extend(['*.xml'])
 
 
+_VERSION_FALLBACK = [
+    ("PKG-INFO", r"^Version: (\d+\.\d+\.\d+.*)$", "{match[1]}"),  # sdist
+    ("NEWS.rst", r"^v(\d+\.\d+\.\d+)", "{match[1]}.dev+{timestamp}"),  # latest version
+]
+
+
+def _extract_version(file: str, pattern: str, fmt: str) -> str | None:
+    with suppress(FileNotFoundError), open(file, encoding="utf-8") as fp:
+        if match := re.search(pattern, fp.read(), re.M):
+            return fmt.format(match=match, timestamp=time.strftime("%Y%m%d"))
+    return None
+
+
+def _get_version() -> str:
+    cmd = ["git", "describe", "--abbrev", "--match", "v?[0-9]*", "--dirty"]
+    try:
+        version = subprocess.check_output(cmd, encoding="utf-8")
+        return _normalization.best_effort_version(version, "{safe}.dev+{sanitized}")
+    except subprocess.CalledProcessError:  # e.g.: git not installed or history missing
+        candidates = filter(None, starmap(_extract_version, _VERSION_FALLBACK))
+        return next(candidates, f"0.0.0.dev+{time.strftime('%Y%m%d')}")
```

Comment:
Reposting from thread in https://github.com/pypa/setuptools/pull/4948/files#r2047494928:

The exact format of the version can be further discussed, e.g. `%Y%m%d` can be changed to more specific time to improve monotonicity... Please let me know if a more specific tag would be better.

The `except` is the code path that takes effect when `git` is not installed and/or tag is not available (tarballs or shallow clones).